src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,559 @@
+  }
+  return mod;
+}
+
+} // end anon
+
+// ModuleBuilder deals with compiling/linking source code to a DLL,
+// and providing the binary DLL as an Array[Byte] which can be broadcast
+// to all workers.
+
+class ModuleBuilder {
+ private:
+  std::string options_;
+  std::string source_;
+  std::string include_;
+  std::string key_;
+  std::string hm_base_;
+  std::string hm_mak_;
+  std::string hm_cpp_;
+  std::string hm_lib_;
+  
+public:
+  ModuleBuilder(
+    const std::string& options,
+    const std::string& source,
+    const std::string& include,
+    const std::string& key
+  ) :
+    options_(options),
+    source_(source),
+    include_(include),
+    key_(key) {
+    // To start with, put dynamic code in $HOME/hail_modules
+    auto base = (config.module_dir_ + "/hm_") + key_;
+    hm_base_ = base;
+    hm_mak_ = (base + config.ext_mak_);
+    hm_cpp_ = (base + config.ext_cpp_);
+    hm_lib_ = (base + config.ext_lib_);
+  }
+  
+  virtual ~ModuleBuilder() { }
+  
+private:
+  void write_cpp() {
+    FILE* f = fopen(hm_cpp_.c_str(), "w");
+    if (!f) { perror("fopen"); return; }
+    fwrite(source_.data(), 1, source_.length(), f);
+    fclose(f);
+  }
+  
+  void write_mak() {
+    FILE* f = fopen(hm_mak_.c_str(), "w");
+    if (!f) { perror("fopen"); return; }
+    fprintf(f, ".PHONY: FORCE\n");
+    fprintf(f, "\n");
+    fprintf(f, "MODULE    := hm_%s\n", key_.c_str());
+    fprintf(f, "MODULE_SO := $(MODULE)%s\n", config.ext_lib_.c_str());
+    fprintf(f, "ifndef JAVA_HOME\n");
+    fprintf(f, "  TMP :=$(shell java -XshowSettings:properties -version 2>&1 | fgrep -i java.home)\n");
+    fprintf(f, "  JAVA_HOME :=$(shell dirname $(filter-out java.home =,$(TMP)))\n");
+    fprintf(f, "endif\n");
+    fprintf(f, "JAVA_INCLUDE :=$(JAVA_HOME)/include\n");
+    fprintf(f, "CXXFLAGS  := \\\n");
+    fprintf(f, "  -std=c++11 -fPIC -march=native -fno-strict-aliasing -Wall \\\n");
+    fprintf(f, "  -I$(JAVA_INCLUDE) \\\n");
+    fprintf(f, "  -I$(JAVA_INCLUDE)/%s \\\n", config.java_md_.c_str());
+    fprintf(f, "  -I%s \\\n", include_.c_str());
+    bool have_oflag = (strstr(options_.c_str(), "-O") != nullptr);
+    fprintf(f, "  %s%s \\\n", have_oflag ? "" : "-O3", options_.c_str());
+    fprintf(f, "  -DHAIL_MODULE=$(MODULE)\n");
+    fprintf(f, "LIBFLAGS := -fvisibility=default %s\n", 
+      config.is_darwin_ ? "-dynamiclib -Wl,-undefined,dynamic_lookup"
+                         : "-rdynamic -shared");
+    fprintf(f, "\n");
+    // build .so from .cpp
+    fprintf(f, "$(MODULE_SO): FORCE\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) -o $(MODULE).o -c $(MODULE).cpp 2> $(MODULE).err\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) $(LIBFLAGS) -o $@ $(MODULE).o 2>> $(MODULE).err\n");
+    fprintf(f, "\n");
+    fclose(f);
+  }
+
+public:
+  bool try_to_build() {
+    // Try to create the .new file, we hold the global lock so there is no race
+    int fd = ::open(hm_lib_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
```

Comment:
I'm confused what this is doing.  Isn't the Makefile going to write hm_lib_?  Also, comment out of date now.